File: Multiplatform/Architecture/src/commonMain/kotlin/StorePlugin.kt
package Multiplatform.Architecture

interface StorePlugin<State, in Action, in Effect> {
    fun onWillDispatch(action: Action, state: State)
    fun onDidDispatch(action: Action, state: State)

    fun onWillReduce(action: Action, state: State)
    fun onDidReduce(action: Action, state: State)

    fun onWillHandleEffect(effect: Effect)
    fun onDidHandleEffect(effect: Effect)
}

File: Mobile/Libraries/iOS/iOSArchitecture/API/StorePlugin.swift
public protocol StorePlugin {
    associatedtype State
    associatedtype Action
    associatedtype Effect

    func onWillDispatch(action: Action, state: State)
    func onDidDispatch(action: Action, state: State)

    func onWillReduce(action: Action, state: State)
    func onDidReduce(action: Action, state: State)

    func onWillHandleEffect(effect: Effect)
    func onDidHandleEffect(effect: Effect)
}